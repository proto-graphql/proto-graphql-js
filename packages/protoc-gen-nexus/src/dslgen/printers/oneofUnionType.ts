import type { Registry } from "@bufbuild/protobuf";
import {
  type OneofUnionType,
  type SquashedOneofUnionType,
  compact,
  protobufGraphQLExtensions,
} from "@proto-graphql/codegen-core";
import { type Code, code, joinCode, literalOf } from "ts-poet";

import {
  type NexusPrinterOptions,
  fieldType,
  impNexus,
  nexusTypeDef,
} from "./util.js";

/**
 * @example
 * ```ts
 * export cosnt Oneof = unionType({
 *   name: "Oneof",
 *   // ...
 * })
 * ```
 */
export function createOneofUnionTypeCode(
  type: OneofUnionType | SquashedOneofUnionType,
  registry: Registry,
  opts: NexusPrinterOptions,
): Code {
  const typeOpts = {
    name: type.typeName,
    description: type.description,
    definition: code`(t) => {
      t.members(${joinCode(
        type.fields.map((f) => fieldType(f, opts)),
        { on: "," },
      )});
    }`,
    extensions: protobufGraphQLExtensions(type, registry),
  };
  return code`export const ${nexusTypeDef(type)} = ${impNexus(
    "unionType",
  )}(${literalOf(compact(typeOpts))});`;
}
