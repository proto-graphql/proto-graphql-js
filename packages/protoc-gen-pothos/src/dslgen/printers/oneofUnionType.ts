import type { Registry } from "@bufbuild/protobuf";
import {
  type OneofUnionType,
  type SquashedOneofUnionType,
  compact,
  protobufGraphQLExtensions,
} from "@proto-graphql/codegen-core";
import { type Code, code, literalOf } from "ts-poet";

import {
  type PothosPrinterOptions,
  fieldTypeRef,
  pothosBuilder,
  pothosRef,
} from "./util";

/**
 * @example
 * ```ts
 * export cosnt Oneof = builder.unionType("Oneof", {
 *   types: [...],
 *   // ...
 * })
 * ```
 */
export function createOneofUnionTypeCode(
  type: OneofUnionType | SquashedOneofUnionType,
  registry: Registry,
  opts: PothosPrinterOptions,
): Code {
  const typeOpts = {
    types: type.fields.map((f) => fieldTypeRef(f, opts)),
    description: type.description,
    extensions: protobufGraphQLExtensions(type, registry),
  };
  return code`
    export const ${pothosRef(type)} =
      ${pothosBuilder(type, opts)}.unionType(${literalOf(
        type.typeName,
      )}, ${literalOf(compact(typeOpts))});
  `;
}
