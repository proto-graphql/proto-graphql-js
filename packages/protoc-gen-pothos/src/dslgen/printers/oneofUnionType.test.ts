import { createFileRegistry } from "@bufbuild/protobuf";
import {
  OneofUnionType,
  SquashedOneofUnionType,
  type TypeOptions,
  defaultScalarMapping,
  defaultScalarMappingForTsProto,
} from "@proto-graphql/codegen-core";
import {
  type TestapisPackage,
  getTestapisFileDescriptorSet,
} from "@proto-graphql/testapis-proto";
import { describe, expect, test } from "vitest";
import { createOneofUnionTypeCode } from "./oneofUnionType.js";
import type { PothosPrinterOptions } from "./util.js";

function generateOneofUnionTypeCode(
  packageName: TestapisPackage,
  typeNameInProto: string,
  oneofFieldName: string,
  options: PothosPrinterOptions,
): string {
  const typeOptions: TypeOptions = {
    partialInputs: false,
    scalarMapping:
      options.protobuf === "ts-proto"
        ? defaultScalarMappingForTsProto
        : defaultScalarMapping,
    ignoreNonMessageOneofFields: false,
  };

  const descSet = getTestapisFileDescriptorSet(packageName);
  const registry = createFileRegistry(descSet);
  const descMsg = registry.getMessage(`${packageName}.${typeNameInProto}`);
  if (descMsg === undefined) {
    throw new Error(
      `Message ${typeNameInProto} not found in package ${packageName}`,
    );
  }

  const descOneof = descMsg.oneofs.find((d) => d.name === oneofFieldName);
  if (descOneof === undefined) {
    throw new Error(
      `Oneof field ${oneofFieldName} not found in message ${typeNameInProto} in package ${packageName}`,
    );
  }

  const oneofType = new OneofUnionType(descOneof, typeOptions);

  const code = createOneofUnionTypeCode(
    oneofType,
    registry,
    options,
  ).toString();

  return code.toString();
}

function generateSquashedOneofUnionTypeCode(
  packageName: TestapisPackage,
  typeNameInProto: string,
  options: PothosPrinterOptions,
): string {
  const typeOptions: TypeOptions = {
    partialInputs: false,
    scalarMapping:
      options.protobuf === "ts-proto"
        ? defaultScalarMappingForTsProto
        : defaultScalarMapping,
    ignoreNonMessageOneofFields: false,
  };

  const descSet = getTestapisFileDescriptorSet(packageName);
  const registry = createFileRegistry(descSet);
  const descMsg = registry.getMessage(`${packageName}.${typeNameInProto}`);
  if (descMsg === undefined) {
    throw new Error(
      `Message ${typeNameInProto} not found in package ${packageName}`,
    );
  }

  const oneofType = new SquashedOneofUnionType(descMsg, typeOptions);

  const code = createOneofUnionTypeCode(
    oneofType,
    registry,
    options,
  ).toString();

  return code.toString();
}

type OneofTestCase = {
  test: string;
  args: {
    packageName: TestapisPackage;
    typeNameInProto: string;
    oneofFieldName: string;
  };
};

type SquashedTestCase = {
  test: string;
  args: {
    packageName: TestapisPackage;
    typeNameInProto: string;
  };
};

type TestCase = OneofTestCase | SquashedTestCase;

type TestSuite = {
  suite: string;
  options: PothosPrinterOptions;
  cases: TestCase[];
};

const testSuites: TestSuite[] = [
  {
    suite: "ts-proto",
    options: {
      dsl: "pothos",
      protobuf: "ts-proto" as const,
      importPrefix: "@testapis/ts-proto",
      emitImportedFiles: false,
      fileLayout: "proto_file",
      filenameSuffix: ".pothos",
      pothos: {
        builderPath: "../../builder",
      },
    },
    cases: [
      {
        test: "generates code for a required oneof union",
        args: {
          packageName: "testapis.oneof",
          typeNameInProto: "OneofParent",
          oneofFieldName: "required_oneof_members",
        },
      } as OneofTestCase,
      {
        test: "generates code for an optional oneof union",
        args: {
          packageName: "testapis.oneof",
          typeNameInProto: "OneofParent",
          oneofFieldName: "optional_oneof_members",
        },
      } as OneofTestCase,
      {
        test: "generates code for a squashed oneof union",
        args: {
          packageName: "testapis.extensions",
          typeNameInProto: "PrefixedMessage.SquashedMessage",
        },
      } as SquashedTestCase,
      {
        test: "generates code for imported oneof member",
        args: {
          packageName: "testapis.edgecases.import_oneof_member_from_other_file",
          typeNameInProto: "OneofParent",
          oneofFieldName: "oneof_field",
        },
      } as OneofTestCase,
    ],
  },
  {
    suite: "protobuf-es",
    options: {
      dsl: "pothos",
      protobuf: "protobuf-es" as const,
      importPrefix: "@testapis/protobuf-es",
      emitImportedFiles: false,
      fileLayout: "proto_file",
      filenameSuffix: ".pothos",
      pothos: {
        builderPath: "../../builder",
      },
    },
    cases: [
      {
        test: "generates code for a required oneof union",
        args: {
          packageName: "testapis.oneof",
          typeNameInProto: "OneofParent",
          oneofFieldName: "required_oneof_members",
        },
      } as OneofTestCase,
      {
        test: "generates code for an optional oneof union",
        args: {
          packageName: "testapis.oneof",
          typeNameInProto: "OneofParent",
          oneofFieldName: "optional_oneof_members",
        },
      } as OneofTestCase,
      {
        test: "generates code for a squashed oneof union",
        args: {
          packageName: "testapis.extensions",
          typeNameInProto: "PrefixedMessage.SquashedMessage",
        },
      } as SquashedTestCase,
    ],
  },
  {
    suite: "with file layout graphql_type",
    options: {
      dsl: "pothos",
      protobuf: "ts-proto" as const,
      importPrefix: "@testapis/ts-proto",
      emitImportedFiles: false,
      fileLayout: "graphql_type",
      filenameSuffix: ".pothos",
      pothos: {
        builderPath: "../../builder",
      },
    },
    cases: [
      {
        test: "generates code with correct imports for graphql_type layout",
        args: {
          packageName: "testapis.oneof",
          typeNameInProto: "OneofParent",
          oneofFieldName: "required_oneof_members",
        },
      } as OneofTestCase,
    ],
  },
];

describe("createOneofUnionTypeCode", () => {
  for (const { suite, options, cases } of testSuites) {
    describe(suite, () => {
      test.each(cases)("$test", ({ args }) => {
        if ("oneofFieldName" in args) {
          const code = generateOneofUnionTypeCode(
            args.packageName,
            args.typeNameInProto,
            args.oneofFieldName,
            options,
          );
          expect(code).toMatchSnapshot();
        } else {
          const code = generateSquashedOneofUnionTypeCode(
            args.packageName,
            args.typeNameInProto,
            options,
          );
          expect(code).toMatchSnapshot();
        }
      });
    });
  }
});