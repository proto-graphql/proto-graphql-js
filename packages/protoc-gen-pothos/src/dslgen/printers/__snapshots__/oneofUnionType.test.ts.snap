// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createOneofUnionTypeCode > protobuf-es > 'generates code for a required oneof u…' 1`] = `
"import { builder } from "../../../../builder";

export const OneofParentRequiredOneofMembers$Ref = builder.unionType("OneofParentRequiredOneofMembers", {
  "types": [OneofMemberMessage1$Ref, OneofMemberMessage2$Ref],
  "description": "Required. disallow not_set.",
  "extensions": {
    "protobufOneof": {
      "fullName": "testapis.oneof.OneofParent.required_oneof_members",
      "name": "required_oneof_members",
      "messageName": "OneofParent",
      "package": "testapis.oneof",
      "fields": [{ "name": "required_message1", "type": "testapis.oneof.OneofMemberMessage1" }, {
        "name": "required_message2",
        "type": "testapis.oneof.OneofMemberMessage2",
      }],
    },
  },
});
"
`;

exports[`createOneofUnionTypeCode > protobuf-es > 'generates code for a squashed oneof u…' 1`] = `
"import { builder } from "../../../../builder";

export const TestPrefixPrefixedMessageSquashedMessage$Ref = builder.unionType(
  "TestPrefixPrefixedMessageSquashedMessage",
  {
    "types": [TestPrefixPrefixedMessageInnerMessage$Ref, TestPrefixPrefixedMessageInnerMessage2$Ref],
    "extensions": {
      "protobufOneof": {
        "fullName": "testapis.extensions.PrefixedMessage.SquashedMessage",
        "name": "SquashedMessage",
        "package": "testapis.extensions",
        "fields": [{
          "name": "oneof_field",
          "type": "testapis.extensions.PrefixedMessage.InnerMessage",
          "options": { "[graphql.object_type]": { "squashUnion": true } },
        }, {
          "name": "oneof_field_2",
          "type": "testapis.extensions.PrefixedMessage.InnerMessage2",
          "options": { "[graphql.object_type]": { "squashUnion": true } },
        }],
      },
    },
  },
);
"
`;

exports[`createOneofUnionTypeCode > protobuf-es > 'generates code for an optional oneof …' 1`] = `
"import { builder } from "../../../../builder";

export const OneofParentOptionalOneofMembers$Ref = builder.unionType("OneofParentOptionalOneofMembers", {
  "types": [OneofMemberMessage1$Ref, OneofMemberMessage2$Ref],
  "extensions": {
    "protobufOneof": {
      "fullName": "testapis.oneof.OneofParent.optional_oneof_members",
      "name": "optional_oneof_members",
      "messageName": "OneofParent",
      "package": "testapis.oneof",
      "fields": [{ "name": "optoinal_message1", "type": "testapis.oneof.OneofMemberMessage1" }, {
        "name": "optoinal_message2",
        "type": "testapis.oneof.OneofMemberMessage2",
      }],
    },
  },
});
"
`;

exports[`createOneofUnionTypeCode > ts-proto > 'generates code for a required oneof u…' 1`] = `
"import { builder } from "../../../../builder";

export const OneofParentRequiredOneofMembers$Ref = builder.unionType("OneofParentRequiredOneofMembers", {
  "types": [OneofMemberMessage1$Ref, OneofMemberMessage2$Ref],
  "description": "Required. disallow not_set.",
  "extensions": {
    "protobufOneof": {
      "fullName": "testapis.oneof.OneofParent.required_oneof_members",
      "name": "required_oneof_members",
      "messageName": "OneofParent",
      "package": "testapis.oneof",
      "fields": [{ "name": "required_message1", "type": "testapis.oneof.OneofMemberMessage1" }, {
        "name": "required_message2",
        "type": "testapis.oneof.OneofMemberMessage2",
      }],
    },
  },
});
"
`;

exports[`createOneofUnionTypeCode > ts-proto > 'generates code for a squashed oneof u…' 1`] = `
"import { builder } from "../../../../builder";

export const TestPrefixPrefixedMessageSquashedMessage$Ref = builder.unionType(
  "TestPrefixPrefixedMessageSquashedMessage",
  {
    "types": [TestPrefixPrefixedMessageInnerMessage$Ref, TestPrefixPrefixedMessageInnerMessage2$Ref],
    "extensions": {
      "protobufOneof": {
        "fullName": "testapis.extensions.PrefixedMessage.SquashedMessage",
        "name": "SquashedMessage",
        "package": "testapis.extensions",
        "fields": [{
          "name": "oneof_field",
          "type": "testapis.extensions.PrefixedMessage.InnerMessage",
          "options": { "[graphql.object_type]": { "squashUnion": true } },
        }, {
          "name": "oneof_field_2",
          "type": "testapis.extensions.PrefixedMessage.InnerMessage2",
          "options": { "[graphql.object_type]": { "squashUnion": true } },
        }],
      },
    },
  },
);
"
`;

exports[`createOneofUnionTypeCode > ts-proto > 'generates code for an optional oneof …' 1`] = `
"import { builder } from "../../../../builder";

export const OneofParentOptionalOneofMembers$Ref = builder.unionType("OneofParentOptionalOneofMembers", {
  "types": [OneofMemberMessage1$Ref, OneofMemberMessage2$Ref],
  "extensions": {
    "protobufOneof": {
      "fullName": "testapis.oneof.OneofParent.optional_oneof_members",
      "name": "optional_oneof_members",
      "messageName": "OneofParent",
      "package": "testapis.oneof",
      "fields": [{ "name": "optoinal_message1", "type": "testapis.oneof.OneofMemberMessage1" }, {
        "name": "optoinal_message2",
        "type": "testapis.oneof.OneofMemberMessage2",
      }],
    },
  },
});
"
`;

exports[`createOneofUnionTypeCode > ts-proto > 'generates code for imported oneof mem…' 1`] = `
"import { builder } from "../../../../../builder";
import { OneofMember1$Ref, OneofMember2$Ref } from "./member.pothos";

export const OneofParentOneofField$Ref = builder.unionType("OneofParentOneofField", {
  "types": [OneofMember1$Ref, OneofMember2$Ref],
  "extensions": {
    "protobufOneof": {
      "fullName": "testapis.edgecases.import_oneof_member_from_other_file.OneofParent.oneof_field",
      "name": "oneof_field",
      "messageName": "OneofParent",
      "package": "testapis.edgecases.import_oneof_member_from_other_file",
      "fields": [{ "name": "member1", "type": "testapis.edgecases.import_oneof_member_from_other_file.OneofMember1" }, {
        "name": "member2",
        "type": "testapis.edgecases.import_oneof_member_from_other_file.OneofMember2",
      }],
    },
  },
});
"
`;

exports[`createOneofUnionTypeCode > with file layout graphql_type > 'generates code with correct imports f…' 1`] = `
"import { builder } from "../../../../builder";
import { OneofMemberMessage1$Ref } from "./OneofMemberMessage1.pothos";
import { OneofMemberMessage2$Ref } from "./OneofMemberMessage2.pothos";

export const OneofParentRequiredOneofMembers$Ref = builder.unionType("OneofParentRequiredOneofMembers", {
  "types": [OneofMemberMessage1$Ref, OneofMemberMessage2$Ref],
  "description": "Required. disallow not_set.",
  "extensions": {
    "protobufOneof": {
      "fullName": "testapis.oneof.OneofParent.required_oneof_members",
      "name": "required_oneof_members",
      "messageName": "OneofParent",
      "package": "testapis.oneof",
      "fields": [{ "name": "required_message1", "type": "testapis.oneof.OneofMemberMessage1" }, {
        "name": "required_message2",
        "type": "testapis.oneof.OneofMemberMessage2",
      }],
    },
  },
});
"
`;
